File: TPs/TP02/Fila.java
// Algoritmos e Estruturas de Dados 2 ----- Trabalho prático 2
// Questão 13 ----- Fila Circular com Alocação Sequencial em Java

import java.io.*;
import java.util.*;

class Data {
    int dia, mes, ano;

    public Data(int d, int m, int a) {
        dia = d; mes = m; ano = a;
    }

    public static Data parseData(String s) {
        return new Data(
            Integer.parseInt(s.substring(8,10)),
            Integer.parseInt(s.substring(5,7)),
            Integer.parseInt(s.substring(0,4))
        );
    }

    public String formatar() {
        return String.format("%02d/%02d/%04d", dia, mes, ano);
    }
}

class Hora {
    int h, m;

    public Hora(int h, int m) {
        this.h = h; this.m = m;
    }

    public static Hora parseHora(String s) {
        return new Hora(
            Integer.parseInt(s.substring(0,2)),
            Integer.parseInt(s.substring(3,5))
        );
    }

    public String formatar() {
        return String.format("%02d:%02d", h, m);
    }
}

class Restaurante {

    int id, capacidade, faixaPreco;
    double avaliacao;
    String nome, cidade;
    String[] tipos;
    Hora abertura, fechamento;
    Data data;
    boolean aberto;

    public int getId() { return id; }
    public String getNome() { return nome; }
    public int getAno() { return data.ano; }

    public static Restaurante parse(String linha) {

        String[] c = linha.split(",");

        Restaurante r = new Restaurante();

        r.id = Integer.parseInt(c[0]);
        r.nome = c[1];
        r.cidade = c[2];
        r.capacidade = Integer.parseInt(c[3]);
        r.avaliacao = Double.parseDouble(c[4]);

        r.tipos = c[5].split(";");
        r.faixaPreco = c[6].length();

        String[] h = c[7].split("-");
        r.abertura = Hora.parseHora(h[0]);
        r.fechamento = Hora.parseHora(h[1]);

        r.data = Data.parseData(c[8]);
        r.aberto = c[9].equals("true");

        return r;
    }

    public String formatar() {

        String tiposStr = "[" + String.join(",", tipos) + "]";
        String preco = "$".repeat(faixaPreco);

        return "[" + id + " ## " + nome + " ## " + cidade + " ## " + capacidade +
               " ## " + avaliacao + " ## " + tiposStr + " ## " + preco +
               " ## " + abertura.formatar() + "-" + fechamento.formatar() +
               " ## " + data.formatar() + " ## " + aberto + "]";
    }
}

class Colecao {

    Restaurante[] array = new Restaurante[10000];
    int n = 0;

    public void lerCSV() throws Exception {

        BufferedReader br = new BufferedReader(new FileReader("/tmp/restaurantes.csv"));
        br.readLine();

        String linha;
        while ((linha = br.readLine()) != null) {
            array[n++] = Restaurante.parse(linha);
        }

        br.close();
    }

    public Restaurante buscar(int id) {
        for (int i = 0; i < n; i++) {
            if (array[i].getId() == id) return array[i];
        }
        return null;
    }
}

// Fila Circular
public class Fila {

    private Restaurante[] array;
    private int primeiro, ultimo;

    public Fila(int tamanho) {
        array = new Restaurante[tamanho + 1]; 
        primeiro = ultimo = 0;
    }

    private int size() {
        return (ultimo - primeiro + array.length) % array.length;
    }

    public void inserir(Restaurante x) {

        // Remover antes, caso esteje cheio
        if ((ultimo + 1) % array.length == primeiro) {
            remover();
        }

        array[ultimo] = x;
        ultimo = (ultimo + 1) % array.length;

        System.out.println("(I)" + media());
    }

    public Restaurante remover() {

        Restaurante r = array[primeiro];
        primeiro = (primeiro + 1) % array.length;

        System.out.println("(R)" + r.getNome());
        return r;
    }

    private int media() {

        int soma = 0, count = size();

        for (int i = primeiro; i != ultimo; i = (i + 1) % array.length) {
            soma += array[i].getAno();
        }

        return Math.round((float)soma / count);
    }

    public void mostrar() {
        for (int i = primeiro; i != ultimo; i = (i + 1) % array.length) {
            System.out.println(array[i].formatar());
        }
    }

    public static void main(String[] args) throws Exception {

        Scanner sc = new Scanner(System.in);

        Colecao c = new Colecao();
        c.lerCSV();

        Fila f = new Fila(5);

        int id;

        while ((id = sc.nextInt()) != -1) {
            f.inserir(c.buscar(id));
        }

        int m = sc.nextInt();
        sc.nextLine();

        for (int i = 0; i < m; i++) {

            String[] p = sc.nextLine().split(" ");

            if (p[0].equals("I")) {
                f.inserir(c.buscar(Integer.parseInt(p[1])));
            } else {
                f.remover();
            }
        }

        f.mostrar();
        sc.close();
    }
}

File: TPs/TP03/Fila.c
//Algoritmos e Estruturas de Dados 2 ----- Trabalho Prático 3
//Questão 07 ----- Fila com Alocação Flexível em C

#include <stdio.h>
#include <stdlib.h>
#include <string.h>

typedef struct Data {
    int ano;
    int mes;
    int dia;
} Data;

typedef struct Hora {
    int hora;
    int minuto;
} Hora;

typedef struct Restaurante {

    int id;

    char nome[200];

    char cidade[100];

    int capacidade;

    double avaliacao;

    char tipos_cozinha[20][100];

    int quantidade_tipos;

    int faixa_preco;

    Hora horario_abertura;

    Hora horario_fechamento;

    Data data_abertura;

    int aberto;

} Restaurante;

typedef struct Celula {

    Restaurante* restaurante;

    struct Celula* prox;

} Celula;

typedef struct Fila {

    Celula* primeiro;

    Celula* ultimo;

} Fila;

typedef struct ColecaoRestaurantes {

    Restaurante* restaurantes[10000];

    int tamanho;

} ColecaoRestaurantes;

Data parse_data(char* s) {

    Data data;

    sscanf(s,
           "%d-%d-%d",
           &data.ano,
           &data.mes,
           &data.dia);

    return data;
}

Hora parse_hora(char* s) {

    Hora hora;

    sscanf(s,
           "%d:%d",
           &hora.hora,
           &hora.minuto);

    return hora;
}

void formatar_data(Data* data,
                   char* buffer) {

    sprintf(buffer,
            "%02d/%02d/%04d",
            data->dia,
            data->mes,
            data->ano);
}

void formatar_hora(Hora* hora,
                   char* buffer) {

    sprintf(buffer,
            "%02d:%02d",
            hora->hora,
            hora->minuto);
}

void separar_csv(char* linha,
                 char campos[][300]) {

    int i = 0;
    int j = 0;
    int campo = 0;
    int aspas = 0;

    while (linha[i] != '\0' &&
           linha[i] != '\n') {

        if (linha[i] == '"') {

            aspas = !aspas;
        }
        else if (linha[i] == ',' &&
                 aspas == 0) {

            campos[campo][j] = '\0';

            campo++;

            j = 0;
        }
        else {

            campos[campo][j] =
                    linha[i];

            j++;
        }

        i++;
    }

    campos[campo][j] = '\0';
}

Restaurante* parse_restaurante(char* linha) {

    Restaurante* restaurante =
            (Restaurante*)
            malloc(sizeof(Restaurante));

    char campos[10][300];

    separar_csv(linha,
                campos);

    restaurante->id =
            atoi(campos[0]);

    strcpy(restaurante->nome,
           campos[1]);

    strcpy(restaurante->cidade,
           campos[2]);

    restaurante->capacidade =
            atoi(campos[3]);

    restaurante->avaliacao =
            atof(campos[4]);

    restaurante->quantidade_tipos = 0;

    int i = 0;
    int j = 0;
    int tipo = 0;

    while (campos[5][i] != '\0') {

        if (campos[5][i] == ';') {

            restaurante->tipos_cozinha[tipo][j] =
                    '\0';

            tipo++;

            j = 0;
        }
        else {

            restaurante->tipos_cozinha[tipo][j] =
                    campos[5][i];

            j++;
        }

        i++;
    }

    restaurante->tipos_cozinha[tipo][j] =
            '\0';

    restaurante->quantidade_tipos =
            tipo + 1;

    restaurante->faixa_preco =
            strlen(campos[6]);

    char abertura[20];
    char fechamento[20];

    sscanf(campos[7],
           "%[^-]-%s",
           abertura,
           fechamento);

    restaurante->horario_abertura =
            parse_hora(abertura);

    restaurante->horario_fechamento =
            parse_hora(fechamento);

    restaurante->data_abertura =
            parse_data(campos[8]);

    restaurante->aberto =
            strstr(campos[9], "true") != NULL ||
            strstr(campos[9], "TRUE") != NULL ||
            strstr(campos[9], "True") != NULL;
    
    return restaurante;
}

void formatar_restaurante(Restaurante* restaurante,
                          char* buffer) {

    char data[30];

    char abertura[20];

    char fechamento[20];

    formatar_data(&restaurante->data_abertura,
                  data);

    formatar_hora(&restaurante->horario_abertura,
                  abertura);

    formatar_hora(&restaurante->horario_fechamento,
                  fechamento);

    char tipos[500] = "[";

    for (int i = 0;
         i < restaurante->quantidade_tipos;
         i++) {

        strcat(tipos,
               restaurante->tipos_cozinha[i]);

        if (i <
            restaurante->quantidade_tipos - 1) {

            strcat(tipos,
                   ",");
        }
    }

    strcat(tipos,
           "]");

    char preco[10] = "";

    for (int i = 0;
         i < restaurante->faixa_preco;
         i++) {

        strcat(preco,
               "$");
    }

    sprintf(buffer,
            "[%d ## %s ## %s ## %d ## %.1lf ## %s ## %s ## %s-%s ## %s ## %s]",
            restaurante->id,
            restaurante->nome,
            restaurante->cidade,
            restaurante->capacidade,
            restaurante->avaliacao,
            tipos,
            preco,
            abertura,
            fechamento,
            data,
            restaurante->aberto ? "true" : "false");
}

void ler_csv_colecao(ColecaoRestaurantes* colecao,
                     char* path) {

    FILE* arquivo =
            fopen(path,
                  "r");

    char linha[1000];

    fgets(linha,
          1000,
          arquivo);

    colecao->tamanho = 0;

    while (fgets(linha,
                 1000,
                 arquivo) != NULL) {

        colecao->restaurantes[colecao->tamanho] =
                parse_restaurante(linha);

        colecao->tamanho++;
    }

    fclose(arquivo);
}

Restaurante* buscar_por_id(ColecaoRestaurantes* colecao,
                           int id) {

    Restaurante* resp = NULL;

    for (int i = 0;
         i < colecao->tamanho;
         i++) {

        if (colecao->restaurantes[i]->id ==
            id) {

            resp =
                    colecao->restaurantes[i];

            i = colecao->tamanho;
        }
    }

    return resp;
}

Celula* nova_celula(Restaurante* restaurante) {

    Celula* nova =
            (Celula*)
            malloc(sizeof(Celula));

    nova->restaurante =
            restaurante;

    nova->prox = NULL;

    return nova;
}

void iniciar_fila(Fila* fila) {

    fila->primeiro =
            nova_celula(NULL);

    fila->ultimo =
            fila->primeiro;
}

void enfileirar(Fila* fila,
                Restaurante* restaurante) {

    fila->ultimo->prox =
            nova_celula(restaurante);

    fila->ultimo =
            fila->ultimo->prox;
}

Restaurante* desenfileirar(Fila* fila) {

    if (fila->primeiro ==
        fila->ultimo) {

        return NULL;
    }

    Celula* temp =
            fila->primeiro->prox;

    Restaurante* resp =
            temp->restaurante;

    fila->primeiro->prox =
            temp->prox;

    if (temp ==
        fila->ultimo) {

        fila->ultimo =
                fila->primeiro;
    }

    free(temp);

    return resp;
}

void mostrar(Fila* fila) {

    Celula* i =
            fila->primeiro->prox;

    char buffer[1000];

    while (i != NULL) {

        formatar_restaurante(
                i->restaurante,
                buffer);

        printf("%s\n",
               buffer);

        i = i->prox;
    }
}

int main() {

    ColecaoRestaurantes colecao;

    ler_csv_colecao(&colecao,
                    "/tmp/restaurantes.csv");

    Fila fila;

    iniciar_fila(&fila);

    int id;

    scanf("%d",
          &id);

    while (id != -1) {

        enfileirar(
                &fila,
                buscar_por_id(
                        &colecao,
                        id));

        scanf("%d",
              &id);
    }

    int n;

    scanf("%d",
          &n);

    getchar();

    for (int i = 0;
         i < n;
         i++) {

        char comando[5];

        scanf("%s",
              comando);

        if (strcmp(comando,
                   "I") == 0) {

            scanf("%d",
                  &id);

            enfileirar(
                    &fila,
                    buscar_por_id(
                            &colecao,
                            id));
        }
        else if (strcmp(comando,
                        "R") == 0) {

            Restaurante* removido =
                    desenfileirar(
                            &fila);

            printf("(R) %s\n",
                   removido->nome);
        }
    }

    mostrar(&fila);

    return 0;
}
